traffic_ops/traffic_ops_golang/deliveryservice/acme.go

```diff
@@ -0,0 +1,363 @@
+		log.Errorf(dsName+": Error getting acme client: %s", err.Error())
+		api.CreateChangeLogRawTx(api.ApiChange, "DS: "+dsName+", ID: "+strconv.Itoa(*dsID)+", ACTION: FAILED to add SSL keys with "+acmeAccount.AcmeProvider, currentUser, logTx)
+		return errors.New("getting acme client: " + err.Error())
+	}
+
+	renewRequest := certificate.Resource{
+		Certificate: []byte(keyObj.Certificate.Crt),
+	}
+
+	cert, err := client.Certificate.Renew(renewRequest, true, false)
+	if err != nil {
+		log.Errorf("Error obtaining acme certificate: %s", err.Error())
+		return err
+	}
+
+	newCertObj := tc.DeliveryServiceSSLKeys{
+		AuthType:        keyObj.AuthType,
+		CDN:             keyObj.CDN,
+		DeliveryService: keyObj.DeliveryService,
+		Key:             keyObj.DeliveryService,
+		Hostname:        keyObj.Hostname,
+		Version:         keyObj.Version + 1,
+	}
+
+	newCertObj.Certificate = tc.DeliveryServiceSSLKeysCertificate{Crt: string(EncodePEMToLegacyPerlRiakFormat(cert.Certificate)), Key: string(EncodePEMToLegacyPerlRiakFormat(cert.PrivateKey)), CSR: string(EncodePEMToLegacyPerlRiakFormat([]byte("ACME Generated")))}
+	if err := riaksvc.PutDeliveryServiceSSLKeysObj(newCertObj, tx, cfg.RiakAuthOptions, cfg.RiakPort); err != nil {
+		log.Errorf("Error posting acme certificate to riak: %s", err.Error())
+		api.CreateChangeLogRawTx(api.ApiChange, "DS: "+dsName+", ID: "+strconv.Itoa(*dsID)+", ACTION: FAILED to add SSL keys with "+acmeAccount.AcmeProvider, currentUser, logTx)
+		return errors.New(dsName + ": putting riak keys: " + err.Error())
+	}
+
+	tx2, err := db.Begin()
+	if err != nil {
+		log.Errorf("starting sql transaction for delivery service " + dsName + ": " + err.Error())
+		return errors.New("starting sql transaction for delivery service " + dsName + ": " + err.Error())
+	}
+
+	if err := updateSSLKeyVersion(dsName, *certVersion+1, tx2); err != nil {
+		log.Errorf("updating SSL key version for delivery service '" + dsName + "': " + err.Error())
+		return errors.New("updating SSL key version for delivery service '" + dsName + "': " + err.Error())
+	}
+	tx2.Commit()
+
+	api.CreateChangeLogRawTx(api.ApiChange, "DS: "+dsName+", ID: "+strconv.Itoa(*dsID)+", ACTION: Added SSL keys with "+acmeAccount.AcmeProvider, currentUser, logTx)
+
+	return nil
+}
+
+func getAcmeAccountConfig(cfg *config.Config, acmeProvider string) *config.ConfigAcmeAccount {
+	for _, acmeCfg := range cfg.AcmeAccounts {
+		if acmeCfg.AcmeProvider == acmeProvider {
+			return &acmeCfg
+		}
+	}
+	return nil
+}
+
+func getDSIdAndVersionFromName(db *sqlx.DB, xmlId string) (*int, *int64, error) {
+	var dsID int
+	var certVersion int64
+
+	if err := db.QueryRow(`SELECT id, ssl_key_version FROM deliveryservice WHERE xml_id = $1`, xmlId).Scan(&dsID, &certVersion); err != nil {
+		return nil, nil, err
+	}
+
+	return &dsID, &certVersion, nil
+}
+
+func GetAcmeClient(acmeAccount *config.ConfigAcmeAccount, userTx *sql.Tx, db *sqlx.DB) (*lego.Client, error) {
+	if acmeAccount.UserEmail == "" {
```

Comment:
Do we need a `nil` check for `acmeAccount` here?

Reply:
i do the nil check before calling this function